File: T1.SqlSharp/T1.SqlSharp/Expressions/ArithmeticOperator.cs
namespace T1.SqlSharp.Expressions;

public enum ArithmeticOperator
{
    Add,
    Subtract,
    Multiply,
    Divide,
    Modulo,
    AddAssign,
    SubtractAssign,
    MultiplyAssign,
    DivideAssign,
    ModuloAssign,
    BitwiseAnd,
    BitwiseOr,
    BitwiseXor,
}

public static class ArithmeticOperatorExtensions
{
    public static string ToSql(this ArithmeticOperator op)
    {
        return op switch
        {
            ArithmeticOperator.Add => "+",
            ArithmeticOperator.Subtract => "-",
            ArithmeticOperator.Multiply => "*",
            ArithmeticOperator.Divide => "/",
            ArithmeticOperator.Modulo => "%",
            ArithmeticOperator.AddAssign => "+=",
            ArithmeticOperator.SubtractAssign => "-=",
            ArithmeticOperator.MultiplyAssign => "*=",
            ArithmeticOperator.DivideAssign => "/=",
            ArithmeticOperator.ModuloAssign => "%=",
            ArithmeticOperator.BitwiseAnd => "&",
            ArithmeticOperator.BitwiseOr => "|",
            ArithmeticOperator.BitwiseXor => "^",
            _ => throw new ArgumentOutOfRangeException(nameof(op), op, null),
        };
    }
    
    public static ArithmeticOperator ToArithmeticOperator(this string op)
    {
        return op switch
        {
            "+" => ArithmeticOperator.Add,
            "-" => ArithmeticOperator.Subtract,
            "*" => ArithmeticOperator.Multiply,
            "/" => ArithmeticOperator.Divide,
            "%" => ArithmeticOperator.Modulo,
            "+=" => ArithmeticOperator.AddAssign,
            "-=" => ArithmeticOperator.SubtractAssign,
            "*=" => ArithmeticOperator.MultiplyAssign,
            "/=" => ArithmeticOperator.DivideAssign,
            "%=" => ArithmeticOperator.ModuloAssign,
            "&" => ArithmeticOperator.BitwiseAnd,
            "|" => ArithmeticOperator.BitwiseOr,
            "^" => ArithmeticOperator.BitwiseXor,
            _ => throw new ArgumentOutOfRangeException(nameof(op), op, null),
        };
    }
}

File: VimSharp/chrome-extension-sharp/src/t1-sqlts/Expressions/ArithmeticOperator.ts
// 對應 C# 的 ArithmeticOperator Enum 與 Extension
export enum ArithmeticOperator {
    Add,
    Subtract,
    Multiply,
    Divide,
    Modulo,
    AddAssign,
    SubtractAssign,
    MultiplyAssign,
    DivideAssign,
    ModuloAssign,
    BitwiseAnd,
    BitwiseOr,
    BitwiseXor,
}

export function arithmeticOperatorToSql(op: ArithmeticOperator): string {
    switch (op) {
        case ArithmeticOperator.Add: return '+';
        case ArithmeticOperator.Subtract: return '-';
        case ArithmeticOperator.Multiply: return '*';
        case ArithmeticOperator.Divide: return '/';
        case ArithmeticOperator.Modulo: return '%';
        case ArithmeticOperator.AddAssign: return '+=';
        case ArithmeticOperator.SubtractAssign: return '-=';
        case ArithmeticOperator.MultiplyAssign: return '*=';
        case ArithmeticOperator.DivideAssign: return '/=';
        case ArithmeticOperator.ModuloAssign: return '%=';
        case ArithmeticOperator.BitwiseAnd: return '&';
        case ArithmeticOperator.BitwiseOr: return '|';
        case ArithmeticOperator.BitwiseXor: return '^';
        default: throw new Error('Unknown ArithmeticOperator: ' + op);
    }
}

export function sqlToArithmeticOperator(op: string): ArithmeticOperator {
    switch (op) {
        case '+': return ArithmeticOperator.Add;
        case '-': return ArithmeticOperator.Subtract;
        case '*': return ArithmeticOperator.Multiply;
        case '/': return ArithmeticOperator.Divide;
        case '%': return ArithmeticOperator.Modulo;
        case '+=': return ArithmeticOperator.AddAssign;
        case '-=': return ArithmeticOperator.SubtractAssign;
        case '*=': return ArithmeticOperator.MultiplyAssign;
        case '/=': return ArithmeticOperator.DivideAssign;
        case '%=': return ArithmeticOperator.ModuloAssign;
        case '&': return ArithmeticOperator.BitwiseAnd;
        case '|': return ArithmeticOperator.BitwiseOr;
        case '^': return ArithmeticOperator.BitwiseXor;
        default: throw new Error('Unknown SQL operator: ' + op);
    }
} 